### PR TITLE
HDDS-11376. Improve ReplicationSupervisor to record replication metrics

### PR DESCRIPTION
**Pull request by jianghuazhu**
## What changes were proposed in this pull request?
When datanode executes ReplicateContainerCommand and ReconstructECContainersCommand, they are all done by ReplicationSupervisor. We should record more detailed metrics.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11376

## How was this patch tested?
When we open the jmx connection of Datanode, we will see the newly added metrics.
Old jmx:
image

New jmx:
image

### CODE REVIEW

**Comment by jianghuazhu** [comment]
ci: https://github.com/jianghuazhu/ozone/actions/runs/10635448262
@fapifta , can you help review this PR?
Thanks.

**Comment by jianghuazhu** [comment]
Thanks @fapifta  for the comment and review.
I will improve `requestCounter` later and record separate metrics for different replication types.
I think timeout and skip should not be treated as failures, which makes it clearer. Otherwise, the metrics recorded by `failureCounter` are easy to confuse.
This is my idea, what do you think? @fapifta

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisorMetrics.java [hunk 1]
```diff
@@ -71,12 +72,51 @@ public void getMetrics(MetricsCollector collector, boolean all) {
         .addGauge(Interns.info("numRequestedReplications",
             "Number of requested replications"),
             supervisor.getReplicationRequestCount())
+        .addGauge(Interns.info("numRequestedECReconstructions",
```
Comment (sodonnel) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisorMetrics.java:75]: If a new task subclass was added in the future (I think reconciliation may be one), then the earlier code in this PR that populates the HashMaps will automatically start gathering metrics for the new class. But the code here will never present them.

Could we add all these gauges by iterating each map, so that if a new task appears in the future, it will automatically get added?

You may need to add some abstract method to AbstractTask to be able to get a meaningful name for each of the tasks that each sub class must implement.

Then instead of `requestCounter.put(task.getClass(), new AtomicLong(0));` have:

```
requestCounter.put(task.getMetricName, new AtomicLong(0));
```

The metric names may be something like:

ECReconstructions
ContainerReplications

etc.

There may be a better way of doing this too - that is the first thought I came up with.

Comment (jianghuazhu) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisorMetrics.java:75]: Thanks @sodonnel  for the comment and review.
I think this is a great idea, I will update it later.

**Comment by jianghuazhu** [comment]
I have updated some of the code.
@sodonnel , can you review again?
Thanks.
ci: https://github.com/jianghuazhu/ozone/actions/runs/10738626305

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/AbstractReplicationTask.java [hunk 2]
```diff
@@ -70,6 +70,14 @@ protected AbstractReplicationTask(long containerID,
     this.term = term;
     queued = Instant.now(clock);
   }
+  
+  protected String getMetricName() {
```
Comment (sodonnel) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/AbstractReplicationTask.java:74]: I would suggest making these abstract methods, which would force any sub-class to implement them or get a compile error. This would reduce the chances of a new type of task of missing these.

**Comment by sodonnel** [comment]
Thanks for updating - I have just one further suggestion commented above, otherwise the changes look good.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/AbstractReplicationTask.java (hunk 2 above)
Comment (jianghuazhu) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/AbstractReplicationTask.java]: Thank you @sodonnel .
I've updated it.

**Review by errose28 — COMMENTED** [review]
Hi @jianghuazhu thanks for working on this. This will be useful for container reconciliation in HDDS-11345 as well. I think we still have to work out the differences in how metrics are tracked in between datanode commands that go through the replication supervisor and those that are handled outside of it. For example in `ReplicateContainerCommandHandler`:
- `totalTime` and `invocationCount` are never assigned. Their related metrics will always be 0.
  - There is no analogous metric for execution times in the supervisor to replace these.
- Queued count is pulled from the supervisor and propagated to the command handler metrics.
  - The supervisor does not expose this metric directly, even though it tracks it.

Basically the way this change is implemented I think it is confusing whether to look for a metric in the replication supervisor or the command handler. IMO all datanode commands should properly fill in a core common set of metrics, and the replication supervisor metrics should extend that with additional data for replication specific commands.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisor.java [hunk 3]
```diff
@@ -438,20 +467,51 @@ public long getMaxReplicationStreams() {
...
   }
 
+  private long getCount(Map<String, AtomicLong> counter) {
+    if (counter.isEmpty()) {
+      return 0;
+    }
+    AtomicLong total = new AtomicLong(0);
+    counter.forEach((key, value) -> {
+      total.set(total.get() + value.get());
+    });
```
[the remark this answers is not in the document]
Comment (jianghuazhu) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisor.java]: Thanks @errose28 .
I have updated it.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisor.java [hunk 4]
```diff
@@ -438,20 +467,49 @@ public long getMaxReplicationStreams() {
     }
   }
 
+  private long getCount(Map<String, AtomicLong> counter) {
+    if (counter.isEmpty()) {
+      return 0;
+    }
+    AtomicLong total = new AtomicLong(0);
```
Comment (sodonnel) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisor.java:474]: This does not need to be an AtomicLong - a simple Long is fine. This is a local variable and we return only its value, so it does not need to be thread safe.

Not particularly important, but you could also remove the `if (counter.isEmpty())` section:

```
long counter = 0;
counter.forEach((key, value) -> total += value.get();
return counter;
```

Comment (jianghuazhu) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisor.java]: Thanks @sodonnel  for your comment and review.
I have updated it, can you review it again?
Thanks.

**Comment by sodonnel** [comment]
> Hi @jianghuazhu thanks for working on this. This will be useful for container reconciliation in [HDDS-11345](https://issues.apache.org/jira/browse/HDDS-11345) as well. I think we still have to work out the differences in how metrics are tracked in between datanode commands that go through the replication supervisor and those that are handled outside of it. For example in `ReplicateContainerCommandHandler`:
> 
>     * `totalTime` and `invocationCount` are never assigned. Their related metrics will always be 0.
>       
>       * There is no analogous metric for execution times in the supervisor to replace these.
> 
>     * Queued count is pulled from the supervisor and propagated to the command handler metrics.
>       
>       * The supervisor does not expose this metric directly, even though it tracks it.
> 
> 
> Basically the way this change is implemented I think it is confusing whether to look for a metric in the replication supervisor or the command handler. IMO all datanode commands should properly fill in a core common set of metrics, and the replication supervisor metrics should extend that with additional data for replication specific commands.

I had a look, and these problems of consistency are pre-existing. I think this PR is an improvement, but there is scope for further changes to bring in the total time and invocation count to the supervisor metrics.

The way the code is structured, the ReplicationContainer command handler simply passes the task to the supervisor, so it cannot track the total time of execution. It could in theory pull the metric from the supervisor on demand, if the supervisor tracked it.

In some respects, replication commands are different to other commands on the DN, so it does somewhat make sense that they have their own metrics from the supervisor.

In summary, I am happy with this change and will give it a +1, but it would make sense to open a Jira and ideally a further PR to add totalTime to the supervisor metrics per command type. That said, I am not really sure how useful that is - all you can get from it is an average, and containers are all different sizes, so its not a very useful metric to analyse the replication performance.

@errose28 - what do you think about proceeding with committing this and having followup work to make things more consistent?

**Review by errose28 — APPROVED** [review]
Yes let's merge this as an improvement over the existing behavior and open a follow-up Jira to fix other gaps in metrics handling across commands.

**Comment by jianghuazhu** [comment]
Thanks @sodonnel  @errose28 .
I will continue to work on HDDS-11444.
